lib/membrane_rtc_engine/endpoints/webrtc/encoding_selector.ex

```diff
@@ -0,0 +1,193 @@
+  end
+
+  @doc """
+  Marks given `encoding` as inactive.
+
+  Returns new selector and encoding to request
+  or `nil` if there are no changes needed.
+  """
+  @spec encoding_inactive(t(), Track.variant()) :: {t(), Track.variant() | nil}
+  def encoding_inactive(selector, encoding) do
+    selector = %__MODULE__{
+      selector
+      | active_encodings: List.delete(selector.active_encodings, encoding)
+    }
+
+    next_encoding = get_next_encoding(selector.active_encodings)
+
+    cond do
+      selector.current_encoding == encoding ->
+        selector = %__MODULE__{selector | current_encoding: nil}
+        select_encoding(selector, next_encoding)
+
+      selector.queued_encoding == encoding and selector.current_encoding == nil ->
+        selector = %__MODULE__{selector | queued_encoding: nil}
+        select_encoding(selector, next_encoding)
+
+      selector.queued_encoding == encoding ->
+        selector = %__MODULE__{selector | queued_encoding: nil}
+        {selector, nil}
+
+      true ->
+        {selector, nil}
+    end
+  end
+
+  @doc """
+  Marks given `encoding` as active.
+
+  See `encoding_inactive/2` for the meaning
+  of return values.
+  """
+  @spec encoding_active(t(), Track.variant()) :: {t(), Track.variant() | nil}
+  def encoding_active(selector, encoding) do
+    selector = %__MODULE__{
+      selector
+      | active_encodings: selector.active_encodings ++ [encoding]
+    }
+
+    next_encoding = get_next_encoding(selector.active_encodings)
+
+    cond do
+      selector.current_encoding == selector.target_encoding ->
+        {selector, nil}
+
+      selector.queued_encoding == selector.target_encoding ->
+        {selector, nil}
+
+      selector.target_encoding == encoding ->
+        select_encoding(selector, encoding)
+
+      true ->
+        select_encoding(selector, next_encoding)
+    end
+  end
+
+  @doc """
+  Sets currently used encoding.
+
+  Should be called when encoding change happens
+  i.e. after receiving `TrackVariantSwitched` event.
+  """
+  @spec current_encoding(t(), Track.variant()) :: t()
+  def current_encoding(%__MODULE__{queued_encoding: encoding} = selector, encoding) do
+    %__MODULE__{selector | current_encoding: encoding, queued_encoding: nil}
+  end
+
+  def current_encoding(%__MODULE__{} = selector, encoding) do
+    %__MODULE__{selector | current_encoding: encoding}
+  end
+
+  @doc """
+  Prioritizes encoding.
+
+  Prioritized encoding will be chosen whenever it
+  is active. See `encoding_inactive/2` for the
+  meaning of return values.
+  """
+  @spec target_encoding(t(), Track.variant()) :: {t(), Track.variant() | nil}
+  def target_encoding(selector, encoding) do
+    cond do
+      encoding not in selector.all_encodings ->
+        Membrane.Logger.warn("""
+        Requested non existing encoding #{inspect(encoding)}. \
+        Available track encodings: #{inspect(selector.all_encodings)}. \
+        Ignoring.\
+        """)
```

Comment:
> Shouldn't that be treated as an error? I.e.
>
> Logged with different level
> Either raising or returning error tuple

:+1: 

> BTW such a check is never done in other functions, so I think it either should be everywhere or nowhere.

`target_encoding` is meant to be called on client request so it can receive invalid input.  `encoding_active`, `encoding_inactive` are always called by ourselves. Maybe this check should be moved one level higher i.e. to the place where we are invoking this function.

Reply:
I agree with the following set of suggestions:

> Maybe this check should be moved one level higher i.e. to the place where we are invoking this function.

> BTW such a check is never done in other functions, so I think it either should be everywhere or nowhere.

> Either raising or returning error tuple

So I'm going to move this check to `TrackReceiver` and raise there. I don't like the logging option, as it doesn't inform the frontend in any way that something went wrong. Because it would be caused by a developer error, I think raise is an acceptable solution.

Reply:
> So I'm going to move this check to TrackReceiver and raise there.

Good call I think